fetching prices : first draft (#220)
Uniswap/docs: SDK_versioned_docs/version-3.0.0/guides/04-fetching-prices.md

---
id: fetching-prices
title: Fetching Spot Prices
---

## Fetching Token Prices with the SDK

> To calculate the spot price of any pool, you can call trivially call `token0Price` or `token1Price` on any Pool object (how nice!). However, in addition to learning a couple of simple steps needed to fetch token prices, you will also learn *how* the SDK calculates these quantities. Going through these calculations will hopefully provide the necessary context behind fixed point numbers, square roots, and difficult-to-understand variable names like `sqrtPriceX96`. :)

Reviewer: would avoid "trivially" because if someone is confused and feeling underwhelmed, it might make them feel dumb.  maybe "simply" instead

Reviewer: if Pool is a class type then should we code-case it?

Reviewer: I don't quite follow the "However" sentence. Am I going to learn about token0Price in this guide? or not? 

maybe:  "This guide will teach you how to fetch the current market price of any token on Uniswap. First you will learn the simple method of calling `token0Price` or `token1Price`. Then you will peek under the hood and learn how the SDK calculates these quantities from <on-chain oracle data?>. Going through these calculations...rest is great"

### Calling the functions

> Similar to other examples, you first must set up your pool. If you’re unsure how to collect all the parameters necessary in creating a Pool instance see [Creating a Pool Instance](https://docs.uniswap.org/sdk/guides/creating-a-pool) or look at this typescript example [here](https://github.com/Uniswap/uniswap-docs/blob/main/sdk-examples/AddAndRemoveLiquidity.tsx). The `Pool` class contains two getter methods `token0Price` and `token1Price` which will return the prices of each token respectively as a `Price`.

Reviewer: don't need 'here' you can put the hyperlink directly on 'example'

After constructing the pool, you can save the token prices as constants:

```typescript
  const DAI_USDC_POOL = new Pool(
    DAI,
    USDC,
    immutables.fee,
    state.sqrtPriceX96.toString(),
    state.liquidity.toString(),
    state.tick
  )
  
  const token0Price =  DAI_USDC_POOL.token0Price
  const token1Price = DAI_USDC_POOL.token1Price
```

### Understanding sqrtPrice

What is `sqrtPriceX96`?

> In Uniswap V3, prices of tokens are stored (rather than derived) to allow pools to perform higher precision operations. In the actual implementation, prices are stored as square roots, hence the `sqrt` prefix. The price is stored as a square root because of the geometric nature of the core AMM algorithm, x*y=k. Essentially, the math works out well when working with the square root of the price. 

Reviewer: where are price stored? in the Pool contract?

Thus, to get a `price` from a `sqrtPriceX96` value, you can execute the following operations:

```python
sqrtPriceX96 = sqrt(price) * 2 ** 96
# divide both sides by 2 ** 96
sqrtPriceX96 / (2 ** 96) = sqrt(price)
# square both sides
(sqrtPriceX96 / (2 ** 96)) ** 2 = price
# expand the squared fraction
(sqrtPriceX96 ** 2) / ((2 ** 96) ** 2)  = price
# multiply the exponents in the denominator to get the final expression
sqrtRatioX96 ** 2 / 2 ** 192 = price
```
You will see that the formula in the last step is how the SDK calculates the prices with the functions [`token0Price`](#token0price) and [`token1Price`](#token1price).

### token0Price

Let's apply the math derived above to the functions `token0Price` and `token1Price`. Note that `sqrtRatioX96` is interchangeable with `sqrtPriceX96`. 

```typescript
  /**
   * Returns the current mid-price of the pool in terms of token0, i.e. the ratio of token1 over token0
   */
  public get token0Price(): Price<Token, Token> {
    return (
      this._token0Price ??
      (this._token0Price = new Price(
        this.token0,
        this.token1,
        Q192,
        JSBI.multiply(this.sqrtRatioX96, this.sqrtRatioX96)
      ))
    )
  }
```

> `token0Price` returns a new `Price` as the ratio of token1 over token0. Note that a `Price` is constructed by :

Reviewer: nit: extra space before the colon

```typescript
constructor(
    baseToken: Token, 
    quoteToken: Token,
    denominator: BigintIsh, 
    numerator: BigintIsh)
```

Let's break down the denominator and the numerator of the returned price and prove that it matches the math derived above. Recall that the expression achieved above is 

```python
price = sqrtRatioX96 ** 2 / 2 ** 192
```

#### The numerator

It's worth noting that the numerator is misleadingly listed *below* the denominator in the constructor for a `Price`. In any case, you will see that the numerator of the fraction is `JSBI.multiply(this.sqrtRatioX96, this.sqrtRatioX96)` which nicely follows the math above: `sqrtPriceX96 ** 2`. 

#### The denominator

> The denominator is `Q192`. To break this number down recall the following constants defined in the SDK :

Reviewer: nit: extra space before colon

### Calling the functions

Similar to other examples, you first must set up your pool. If you’re unsure how to collect all the parameters necessary in creating a Pool instance see [Creating a Pool Instance](https://docs.uniswap.org/sdk/guides/creating-a-pool) or look at this typescript example [here](https://github.com/Uniswap/uniswap-docs/blob/main/sdk-examples/AddAndRemoveLiquidity.tsx). The `Pool` class contains two getter methods `token0Price` and `token1Price` which will return the prices of each token respectively as a `Price`.

After constructing the pool, you can save the token prices as constants:

```typescript
  const DAI_USDC_POOL = new Pool(
    DAI,
    USDC,
    immutables.fee,
    state.sqrtPriceX96.toString(),
    state.liquidity.toString(),
    state.tick
  )
  
  const token0Price =  DAI_USDC_POOL.token0Price
  const token1Price = DAI_USDC_POOL.token1Price
```

### Understanding sqrtPrice

What is `sqrtPriceX96`?

> In Uniswap V3, prices of tokens are stored (rather than derived) to allow pools to perform higher precision operations. In the actual implementation, prices are stored as square roots, hence the variable name `sqrtPriceX96`. The price is stored as a square root because of the geometric nature of the core AMM algorithm, x*y=k. Essentially, the math works out well when working with the square root of the price. 

Reviewer 2: maybe change to "...hence the `sqrt` prefix."

could also add a link to the whitepaper as a reference for the math working out well with square roots.

> In addition, you'll notice the `X96` tacked onto the end of the variable name. This naming convention indicates that it is a [fixed point number](https://en.wikipedia.org/wiki/Fixed-point_arithmetic) with 96 bits of precision preceding the decimal point. Because prices are fractions, using fixed point numbers is a way to represent fractions in integer-only environments and still achieve high precision.

Reviewer 2: ```suggestion
In addition, you'll notice the `X96` suffix at the end of the variable name. This `X*` naming convention is used throughout the Uniswap V3 codebase to indicate values which are encoded as binary [fixed-point numbers](https://en.wikipedia.org/wiki/Fixed-point_arithmetic). Fixed-point is excellent at representing fractions while maintaining consistent fidelity and high precision in integer-only environments like the EVM, making it a perfect fit for representing prices, which of course are ultimately fractions. The number after `X` indicates the number of _fraction bits_ - 96 in this case - reserved for encoding the value after the decimal point. The number of integer bits can be trivially derived from the size of the variable and the number of fraction bits. In this case, `sqrtPriceX96` is stored as a `uint160`, meaning that there are `160 - 96 = 64` integer bits.
```

> An example of calculating a  `sqrtPriceX96` is :

Reviewer 2: ```suggestion
Consider the following derivation, which formalizes the definitions above:
```

```python

> price = 100

Reviewer 2: i'd nix the price line, it's not used elsewhere

sqrtPriceX96 = sqrt(price) * 2 ** 96
```

> Thus, to get the price from the sqrtPriceX96 variable, you can execute the following operations to arrive at a final expression:

Reviewer 2: ```suggestion
Thus, to get a `price` from a `sqrtPriceX96` value, you can execute the following operations:
```

```python
sqrtPriceX96 = sqrt(price) * 2 ** 96

> # divide both sides

Reviewer 2: ```suggestion
# divide both sides by 2 ** 96
```

sqrtPriceX96 / (2 ** 96) = sqrt(price)
# square both sides
(sqrtPriceX96 / (2 ** 96)) ** 2 = price

> # multiply out the square
> (sqrtPriceX96 / (2 ** 96)) * (sqrtPriceX96 / (2 ** 96))  = price
> # square the numerator and denominator
> (sqrtPriceX96 ** 2) / ((2 ** 96) * (2 ** 96)) = price
> # add the exponents in the denominator to get the final expression:
> sqrtRatioX96 ** 2 / 2 ** 192 = price

Reviewer 2: ```suggestion
# expand the squared fraction
(sqrtPriceX96 ** 2) / ((2 ** 96) ** 2)  = price
# multiply the exponents in the denominator to get the final expression
sqrtRatioX96 ** 2 / 2 ** 192 = price
```
